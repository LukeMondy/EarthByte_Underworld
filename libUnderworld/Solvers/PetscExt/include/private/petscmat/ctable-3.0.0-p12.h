/* Contributed by - Mark Adams */

#if !defined(__CTABLE300_H)
#define __CTABLE300_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
