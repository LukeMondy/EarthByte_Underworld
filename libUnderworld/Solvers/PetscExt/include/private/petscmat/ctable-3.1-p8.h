/* Contributed by - Mark Adams */

#if !defined(__CTABLE31_H)
#define __CTABLE31_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include "petscctable.h"
#endif
