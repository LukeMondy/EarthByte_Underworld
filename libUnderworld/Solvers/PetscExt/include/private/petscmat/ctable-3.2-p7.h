/* Contributed by - Mark Adams */

#if !defined(__CTABLE_H)
#define __CTABLE_H

struct _n_PetscTable {
  PetscInt *keytable;
  PetscInt *table;
  PetscInt count;
  PetscInt tablesize;
  PetscInt head;
};

#include <petscctable.h>
#endif
