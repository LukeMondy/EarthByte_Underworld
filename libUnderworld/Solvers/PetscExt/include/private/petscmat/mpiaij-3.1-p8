
#if !defined(__MPIAIJ_H)
#define __MPIAIJ_H

#include "../src/mat/impls/aij/seq/aij.h"
#include "../src/sys/ctable.h"

typedef struct {
  Mat           A,B;                   /* local submatrices: A (diag part),
                                           B (off-diag part) */
  PetscMPIInt   size;                   /* size of communicator */
  PetscMPIInt   rank;                   /* rank of proc in communicator */ 

  /* The following variables are used for matrix assembly */

  PetscTruth    donotstash;             /* PETSC_TRUE if off processor entries dropped */
  MPI_Request   *send_waits;            /* array of send requests */
  MPI_Request   *recv_waits;            /* array of receive requests */
  PetscInt      nsends,nrecvs;         /* numbers of sends and receives */
  PetscScalar   *svalues,*rvalues;     /* sending and receiving data */
  PetscInt      rmax;                   /* maximum message length */
#if defined (PETSC_USE_CTABLE)
  PetscTable    colmap;
#else
  PetscInt      *colmap;                /* local col number of off-diag col */
#endif
  PetscInt      *garray;                /* global index of all off-processor columns */

  /* The following variables are used for matrix-vector products */

  Vec           lvec;              /* local vector */
  Vec           diag;
  VecScatter    Mvctx;             /* scatter context for vector */
  PetscTruth    roworiented;       /* if true, row-oriented input, default true */

  /* The following variables are for MatGetRow() */

  PetscInt      *rowindices;       /* column indices for row */
  PetscScalar   *rowvalues;        /* nonzero values in row */
  PetscTruth    getrowactive;      /* indicates MatGetRow(), not restored */

  /* Used by MatDistribute_MPIAIJ() to allow reuse of previous matrix allocation  and nonzero pattern */
  PetscInt      *ld;               /* number of entries per row left of diagona block */
} Mat_MPIAIJ;

typedef struct { /* used by MatMatMult_MPIAIJ_MPIAIJ and MatPtAP_MPIAIJ_MPIAIJ for reusing symbolic mat product */
  PetscInt       *startsj,*startsj_r;
  PetscScalar    *bufa;
  IS             isrowa,isrowb,iscolb; 
  Mat            *aseq,*bseq,C_seq; /* A_seq=aseq[0], B_seq=bseq[0] */
  Mat            A_loc,B_seq;
  Mat            B_loc,B_oth;  /* partial B_seq -- intend to replace B_seq */
  PetscInt       brstart; /* starting owned rows of B in matrix bseq[0]; brend = brstart+B->m */
  PetscInt       *abi,*abj; /* symbolic i and j arrays of the local product A_loc*B_seq */
  PetscInt       abnz_max;  /* max(abi[i+1] - abi[i]), max num of nnz in a row of A_loc*B_seq */
  MatReuse       reuse; 
  PetscErrorCode (*MatDestroy)(Mat);
  PetscErrorCode (*MatDuplicate)(Mat,MatDuplicateOption,Mat*);
} Mat_MatMatMultMPI;

typedef struct { /* used by MatMerge_SeqsToMPI for reusing the merged matrix */
  PetscLayout    rowmap;
  PetscInt       **buf_ri,**buf_rj;
  PetscMPIInt    *len_s,*len_r,*id_r; /* array of length of comm->size, store send/recv matrix values */
  PetscMPIInt    nsend,nrecv;  
  PetscInt       *bi,*bj; /* i and j array of the local portion of mpi C=P^T*A*P */
  PetscInt       *owners_co,*coi,*coj; /* i and j array of (p->B)^T*A*P - used in the communication */
  PetscErrorCode (*MatDestroy)(Mat);
  PetscErrorCode (*MatDuplicate)(Mat,MatDuplicateOption,Mat*);
} Mat_Merge_SeqsToMPI; 

EXTERN PetscErrorCode MatSetColoring_MPIAIJ(Mat,ISColoring);
EXTERN PetscErrorCode MatSetValuesAdic_MPIAIJ(Mat,void*);
EXTERN PetscErrorCode MatSetValuesAdifor_MPIAIJ(Mat,PetscInt,void*);
EXTERN PetscErrorCode MatSetUpMultiply_MPIAIJ(Mat);
EXTERN PetscErrorCode DisAssemble_MPIAIJ(Mat);
EXTERN PetscErrorCode MatDuplicate_MPIAIJ(Mat,MatDuplicateOption,Mat *);
EXTERN PetscErrorCode MatIncreaseOverlap_MPIAIJ(Mat,PetscInt,IS [],PetscInt);
EXTERN PetscErrorCode MatFDColoringCreate_MPIAIJ(Mat,ISColoring,MatFDColoring);
EXTERN PetscErrorCode MatGetSubMatrices_MPIAIJ (Mat,PetscInt,const IS[],const IS[],MatReuse,Mat *[]);
EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ_All(Mat,MatGetSubMatrixOption,MatReuse,Mat *[]);

EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ(Mat,IS,IS,MatReuse,Mat *);
EXTERN PetscErrorCode MatGetSubMatrix_MPIAIJ_Private (Mat,IS,IS,PetscInt,MatReuse,Mat *);
EXTERN PetscErrorCode MatLoad_MPIAIJ(PetscViewer, const MatType,Mat*);
EXTERN PetscErrorCode MatMatMult_MPIAIJ_MPIAIJ(Mat,Mat,MatReuse,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultSymbolic_MPIAIJ_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatMatMultNumeric_MPIAIJ_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatPtAPSymbolic_MPIAIJ_MPIAIJ(Mat,Mat,PetscReal,Mat*);
EXTERN PetscErrorCode MatPtAPNumeric_MPIAIJ_MPIAIJ(Mat,Mat,Mat);
EXTERN PetscErrorCode MatSetValues_MPIAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[],const PetscScalar [],InsertMode);
EXTERN PetscErrorCode MatDestroy_MPIAIJ_MatMatMult(Mat);
EXTERN PetscErrorCode PetscContainerDestroy_Mat_MatMatMultMPI(void*);
EXTERN PetscErrorCode MatGetRedundantMatrix_MPIAIJ(Mat,PetscInt,MPI_Comm,PetscInt,MatReuse,Mat*);
EXTERN PetscErrorCode MatGetSeqNonzeroStructure_MPIAIJ(Mat,Mat*);

EXTERN_C_BEGIN
EXTERN PetscErrorCode MatMPIAIJSetPreallocation_MPIAIJ(Mat,PetscInt,const PetscInt[],PetscInt,const PetscInt[]);
EXTERN_C_END

#if !defined(PETSC_USE_COMPLEX) && !defined(PETSC_USE_SCALAR_SINGLE) && !defined(PETSC_USE_SCALAR_MAT_SINGLE)
EXTERN PetscErrorCode MatLUFactorSymbolic_MPIAIJ_TFS(Mat,IS,IS,const MatFactorInfo*,Mat*);
#endif
EXTERN PetscErrorCode MatSolve_MPIAIJ(Mat,Vec,Vec);
EXTERN PetscErrorCode MatILUFactor_MPIAIJ(Mat,IS,IS,const MatFactorInfo *);

EXTERN_C_BEGIN
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatGetDiagonalBlock_MPIAIJ(Mat,PetscTruth *,MatReuse,Mat *);
EXTERN PetscErrorCode PETSCMAT_DLLEXPORT MatDiagonalScaleLocal_MPIAIJ(Mat,Vec);

EXTERN_C_END

#endif
