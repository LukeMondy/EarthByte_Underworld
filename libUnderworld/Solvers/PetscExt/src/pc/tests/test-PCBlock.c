/*@ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 **
 **    Copyright (c) 2007 - 2008, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        School of Mathematical Sciences, 
 **        Monash University
 **        Victoria, 3800
 **        Australia
 **
 **    Copyright (c) 2009 - 2010, 
 **        Dave A. May [dave.mayhem23@gmail.com]
 **        Geophysical Fluid Dynamics, 
 **        Department of Earth Sciences,
 **        ETH ZÃ¼rich,
 **        Switzerland
 **
 **    Project:       PetscExt_v3.1
 **    Description:   A collection of extensions to the PETSc library.
 **
 **    All rights reserved.
 **    Redistribution and use in source and binary forms, with or without modification,
 **    are permitted provided that the following conditions are met:
 **
 **        * Redistributions of source code must retain the above copyright notice,
 **              this list of conditions and the following disclaimer.
 **        * Redistributions in binary form must reproduce the above copyright
 **              notice, this list of conditions and the following disclaimer in the
 **              documentation and/or other materials provided with the distribution.
 **        * Neither the name of the Monash University nor the names of its contributors
 **              may be used to endorse or promote products derived from this software
 **              without specific prior written permission.
 **
 **    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 **    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 **    THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 **    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
 **    BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 **    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 **    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 **    HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 **    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT 
 **    OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 **
 **
 **    $Id$
 **
 ** ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~@*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdarg.h>

#include <petsc.h>
#include <petscvec.h>
#include <petscmat.h>
#include <petscksp.h>
#include <petscpc.h>


#include "petscext.h"
#include "petscext_vec.h"
#include "petscext_mat.h"
#include "petscext_pc.h"


void test_singleton( void )
{
	Mat A11, Amat;
	PC pc;
	Vec x,y, bX, bY;
	int n, i,j;
	
	
	PetscPrintf( PETSC_COMM_WORLD, "%s \n", __func__ );
	
	n = 10;
	
	/* Create matrices */
	// A11
	MatCreate( PETSC_COMM_WORLD, &A11 );
	MatSetSizes( A11, PETSC_DECIDE, PETSC_DECIDE, n, n );
	MatSetType( A11, MATSEQAIJ );
	MatSeqAIJSetPreallocation( A11, n, PETSC_NULL );
	
	for( i=0; i<n; i++ ) {
		/*
		for( j=0; j<n; j++ ) {
			MatSetValue( A11, i,j, (double)(10.0 + i), INSERT_VALUES );
		}
		*/
		MatSetValue( A11, i,i, (double)(10.0 + i), INSERT_VALUES );
	}
	MatAssemblyBegin( A11, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A11, MAT_FINAL_ASSEMBLY);
	
	/* Create vectors */
	MatGetVecs( A11, &x, &y );
	
	VecSet( x, 1.0 );
	VecSet( y, 0.0 );
	
	
	/* Create a block matrix */
	MatCreate( PETSC_COMM_WORLD, &Amat );
	MatSetSizes_Block( Amat, 1,1, 1,1 );
	MatSetType( Amat, "block" );
	MatBlockSetValue( Amat, 0,0, A11, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatAssemblyBegin( Amat, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( Amat, MAT_FINAL_ASSEMBLY);
	
	/* Create block vectors */
	VecCreate( PETSC_COMM_WORLD, &bX );
	VecSetSizes( bX, 1,1 );
	VecSetType( bX, "block" );
	VecBlockSetValue( bX, 0, x, INSERT_VALUES );	Stg_VecDestroy( & x );
	VecAssemblyBegin( bX );
	VecAssemblyEnd( bX );
	
	VecCreate( PETSC_COMM_WORLD, &bY );
	VecSetSizes( bY, 1,1 );
	VecSetType( bY, "block" );
	VecBlockSetValue( bY, 0, y, INSERT_VALUES );	Stg_VecDestroy( & y );
	VecAssemblyBegin( bY );
	VecAssemblyEnd( bY );
	
	
	PCCreate( PETSC_COMM_WORLD, &pc );
	PCSetOperators( pc, Amat, Amat, SAME_NONZERO_PATTERN );
	PCSetType( pc, "block" );
	
	
	PCApply( pc, bX, bY );
	//PCView( pc, PETSC_VIEWER_STDOUT_WORLD );
	
	VecView( y, PETSC_VIEWER_STDOUT_WORLD );
	
	
	Stg_PCDestroy( pc );
	
	Stg_VecDestroy( & bX );
	Stg_VecDestroy( & bY );
	
	Stg_MatDestroy( Amat );
	Stg_MatDestroy( A11 );
}


void test_2block_DIAG( void )
{
	Mat A11, A22, Amat;
	PC pc, _pc_i;
	Vec x1,x2,y1,y2, bX, bY;
	int n, n22, i,j;
	const char *bt;
	const PCType pc_type;
	KSP _ksp_i;	

	PetscPrintf( PETSC_COMM_WORLD, "%s \n", __func__ );
	
	n = 10;
	n22 = 12;
	
	/* Create matrices */
	// A11
	MatCreate( PETSC_COMM_WORLD, &A11 );
	MatSetSizes( A11, PETSC_DECIDE, PETSC_DECIDE, n, n );
	MatSetType( A11, MATSEQAIJ );
	MatSeqAIJSetPreallocation( A11, n, PETSC_NULL );
	
	for( i=0; i<n; i++ ) {
		MatSetValue( A11, i,i, (double)(10.0 + i), INSERT_VALUES );
	}
	MatAssemblyBegin( A11, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A11, MAT_FINAL_ASSEMBLY);
	
	// A22
	MatCreate( PETSC_COMM_WORLD, &A22 );
	MatSetSizes( A22, PETSC_DECIDE, PETSC_DECIDE, n22, n22 );
	MatSetType( A22, MATSEQAIJ );
	
	for( i=0; i<n22; i++ ) {
		MatSetValue( A22, i,i, (double)(10.0 + i), INSERT_VALUES );
	}
	MatAssemblyBegin( A22, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A22, MAT_FINAL_ASSEMBLY);
	
	
	
	/* Create vectors */
	MatGetVecs( A11, PETSC_NULL, &x1 );
	MatGetVecs( A22, PETSC_NULL, &x2 );
	
	VecDuplicate( x1, &y1 );
	VecDuplicate( x2, &y2 );
	
	
	VecSet( x1, 1.0 );
	VecSet( x2, 1.0 );
	
	VecSet( y1, 0.0 );
	VecSet( y2, 0.0 );
	
	
	/* Create a block matrix */
	MatCreate( PETSC_COMM_WORLD, &Amat );
	MatSetSizes_Block( Amat, 2,2, 2,2 );
	MatSetType( Amat, "block" );
	MatBlockSetValue( Amat, 0,0, A11, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( Amat, 1,1, A22, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatAssemblyBegin( Amat, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( Amat, MAT_FINAL_ASSEMBLY);
	
	/* Create block vectors */
	VecCreate( PETSC_COMM_WORLD, &bX );
	VecSetSizes( bX, 2,2 );
	VecSetType( bX, "block" );
	
	VecBlockSetValue( bX, 0, x1, INSERT_VALUES );	Stg_VecDestroy( & x1 );
	VecBlockSetValue( bX, 1, x2, INSERT_VALUES );	Stg_VecDestroy( & x2 );
	
	VecAssemblyBegin( bX );
	VecAssemblyEnd( bX );
	
	VecCreate( PETSC_COMM_WORLD, &bY );
	VecSetSizes( bY, 2,2 );
	VecSetType( bY, "block" );
	
	VecBlockSetValue( bY, 0, y1, INSERT_VALUES );	Stg_VecDestroy( & y1 );
	VecBlockSetValue( bY, 1, y2, INSERT_VALUES );	Stg_VecDestroy( & y2 );
	
	VecAssemblyBegin( bY );
	VecAssemblyEnd( bY );
	
	
	PCCreate( PETSC_COMM_WORLD, &pc );
	PCSetOperators( pc, Amat, Amat, SAME_NONZERO_PATTERN );
	PCSetType( pc, "block" );
	
	PCBlockGetSubKSP( pc, 0, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCSetType( _pc_i, "lu" );
	PCBlockGetSubKSP( pc, 1, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCSetType( _pc_i, "jacobi" );
	
	
	PCApply( pc, bX, bY );
	//PCView( pc, PETSC_VIEWER_STDOUT_WORLD );
	
	
	PCBlockGetBlockType( pc, &bt );
	PetscPrintf( PETSC_COMM_WORLD, "block_type: %s \n", bt );
	PCBlockGetSubKSP( pc, 0, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCGetType( _pc_i, &pc_type );
	PetscPrintf( PETSC_COMM_WORLD, "(0,0) pc_type: %s \n", pc_type );
	
	PCBlockGetSubKSP( pc, 1, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCGetType( _pc_i, &pc_type );
	PetscPrintf( PETSC_COMM_WORLD, "(1,1) pc_type: %s \n", pc_type );
	
	
	PetscPrintf( PETSC_COMM_WORLD, "y1 = \n");
	VecView( y1, PETSC_VIEWER_STDOUT_WORLD );
	PetscPrintf( PETSC_COMM_WORLD, "y2 = \n");
	VecView( y2, PETSC_VIEWER_STDOUT_WORLD );
	
	
	Stg_PCDestroy( pc );
	
	Stg_VecDestroy( & bX );
	Stg_VecDestroy( & bY );
	Stg_MatDestroy( Amat );
	
	
	Stg_MatDestroy( A11 );
	Stg_MatDestroy( A22 );
}



void test_2block_UPPER( void )
{
	Mat A11, A22, A12, Amat;
	PC pc, _pc_i;
	Vec x1,x2,y1,y2, bX, bY;
	int n, n22, i,j;
	const char *bt;
	const PCType pc_type;
	KSP _ksp_i;	
	
	PetscPrintf( PETSC_COMM_WORLD, "%s \n", __func__ );
	
	
	n = 10;
	n22 = 12;
	
	/* Create matrices */
	// A11
	MatCreate( PETSC_COMM_WORLD, &A11 );
	MatSetSizes( A11, PETSC_DECIDE, PETSC_DECIDE, n, n );
	MatSetType( A11, MATSEQAIJ );
	MatSeqAIJSetPreallocation( A11, n, PETSC_NULL );
	
	for( i=0; i<n; i++ ) {
		MatSetValue( A11, i,i, (double)(10.0 + i), INSERT_VALUES );
	}
	MatAssemblyBegin( A11, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A11, MAT_FINAL_ASSEMBLY);
	
	// A22
	MatCreate( PETSC_COMM_WORLD, &A22 );
	MatSetSizes( A22, PETSC_DECIDE, PETSC_DECIDE, n22, n22 );
	MatSetType( A22, MATSEQAIJ );
	
	for( i=0; i<n22; i++ ) {
		MatSetValue( A22, i,i, (double)(10.0 + i), INSERT_VALUES );
	}
	MatAssemblyBegin( A22, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A22, MAT_FINAL_ASSEMBLY);
	
	
	// A12
	MatCreate( PETSC_COMM_WORLD, &A12 );
	MatSetSizes( A12, PETSC_DECIDE, PETSC_DECIDE, n, n22 );
	MatSetType( A12, MATSEQAIJ );
	
	for( i=0; i<n; i++ ) {
		for( j=0; j<n22; j++ ) {
			MatSetValue( A12, i,j, (double)(i+j*n), INSERT_VALUES );
	}}
	MatAssemblyBegin( A12, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( A12, MAT_FINAL_ASSEMBLY);
	
	
	
	
	/* Create vectors */
	MatGetVecs( A11, PETSC_NULL, &x1 );
	MatGetVecs( A22, PETSC_NULL, &x2 );
	
	VecDuplicate( x1, &y1 );
	VecDuplicate( x2, &y2 );
	
	
	VecSet( x1, 1.0 );
	VecSet( x2, 1.0 );
	
	VecSet( y1, 0.0 );
	VecSet( y2, 0.0 );
	
	
	/* Create a block matrix */
	MatCreate( PETSC_COMM_WORLD, &Amat );
	MatSetSizes_Block( Amat, 2,2, 2,2 );
	MatSetType( Amat, "block" );
	
	MatBlockSetValue( Amat, 0,0, A11, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( Amat, 0,1, A12, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	MatBlockSetValue( Amat, 1,1, A22, DIFFERENT_NONZERO_PATTERN, INSERT_VALUES );
	
	MatAssemblyBegin( Amat, MAT_FINAL_ASSEMBLY);
	MatAssemblyEnd( Amat, MAT_FINAL_ASSEMBLY);
	
	/* Create block vectors */
	VecCreate( PETSC_COMM_WORLD, &bX );
	VecSetSizes( bX, 2,2 );
	VecSetType( bX, "block" );
	
	VecBlockSetValue( bX, 0, x1, INSERT_VALUES );	Stg_VecDestroy( & x1 );
	VecBlockSetValue( bX, 1, x2, INSERT_VALUES );	Stg_VecDestroy( & x2 );
	
	VecAssemblyBegin( bX );
	VecAssemblyEnd( bX );
	
	VecCreate( PETSC_COMM_WORLD, &bY );
	VecSetSizes( bY, 2,2 );
	VecSetType( bY, "block" );
	
	VecBlockSetValue( bY, 0, y1, INSERT_VALUES );	Stg_VecDestroy( & y1 );
	VecBlockSetValue( bY, 1, y2, INSERT_VALUES );	Stg_VecDestroy( & y2 );
	
	VecAssemblyBegin( bY );
	VecAssemblyEnd( bY );
	
	
	PCCreate( PETSC_COMM_WORLD, &pc );
	PCSetOperators( pc, Amat, Amat, SAME_NONZERO_PATTERN );
	PCSetType( pc, "block" );
	PCBlockSetBlockType( pc, PC_BLOCK_UPPER );
	
	PCApply( pc, bX, bY );
//	PCView( pc, PETSC_VIEWER_STDOUT_WORLD );
	
	PCBlockGetBlockType( pc, &bt );
	PetscPrintf( PETSC_COMM_WORLD, "block_type: %s \n", bt );
	PCBlockGetSubKSP( pc, 0, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCGetType( _pc_i, &pc_type );
	PetscPrintf( PETSC_COMM_WORLD, "(0,0) pc_type: %s \n", pc_type );
	
	PCBlockGetSubKSP( pc, 1, &_ksp_i );
	KSPGetPC( _ksp_i, &_pc_i );
	PCGetType( _pc_i, &pc_type );
	PetscPrintf( PETSC_COMM_WORLD, "(1,1) pc_type: %s \n", pc_type );
	
	PetscPrintf( PETSC_COMM_WORLD, "y1 = \n");
	VecView( y1, PETSC_VIEWER_STDOUT_WORLD );
	PetscPrintf( PETSC_COMM_WORLD, "y2 = \n");
	VecView( y2, PETSC_VIEWER_STDOUT_WORLD );

	
	Stg_PCDestroy( pc );
	
	
	Stg_VecDestroy( & bX );
	Stg_VecDestroy( & bY );
	Stg_MatDestroy( Amat );
	
	
	Stg_MatDestroy( A11 );
	Stg_MatDestroy( A12 );
	Stg_MatDestroy( A22 );
}



int main( int argc, char **args )
{
	int i;
	int BIG=10000;
	PetscTruth flg, val;
	
	
	PetscInitialize( &argc, &args,(char *)0, PETSC_NULL );
	PetscExtVecRegisterAll();
	PetscExtMatRegisterAll();
	PetscExtPCRegisterAll();
	
	
	
	test_singleton();
	test_2block_DIAG();
	test_2block_UPPER();
//	test_2block_LOWER();
	
	PetscOptionsGetTruth( PETSC_NULL, "-leak", &val, &flg );
	if( flg == PETSC_TRUE ) {
		PetscPrintf( PETSC_COMM_WORLD, "  Performing leak test \n");
		for( i=0; i<BIG; i++ ) {
			
		//	test_singleton();
			test_2block_DIAG();
			
			if( i % (BIG/10) == 0 ) {
				PetscPrintf( PETSC_COMM_WORLD, "     Tests conducted: %d/%d \n", i,BIG );
			}
		}
	}
	
	
	
	PetscExtVecRegisterDestroyAll();
	PetscExtMatRegisterDestroyAll();
	PetscExtPCRegisterDestroyAll();
	PetscFinalize();
	return 0;
}
