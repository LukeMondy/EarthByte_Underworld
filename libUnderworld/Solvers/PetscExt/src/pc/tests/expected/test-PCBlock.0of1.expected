test_singleton 
0.1
0.0909091
0.0833333
0.0769231
0.0714286
0.0666667
0.0625
0.0588235
0.0555556
0.0526316
test_2block_DIAG 
block_type: diagonal 
(0,0) pc_type: lu 
(1,1) pc_type: jacobi 
y1 = 
0.1
0.0909091
0.0833333
0.0769231
0.0714286
0.0666667
0.0625
0.0588235
0.0555556
0.0526316
y2 = 
0.1
0.0909091
0.0833333
0.0769231
0.0714286
0.0666667
0.0625
0.0588235
0.0555556
0.0526316
0.05
0.047619
test_2block_UPPER 
block_type: upper 
(0,0) pc_type: ilu 
(1,1) pc_type: ilu 
y1 = 
-3.7361
-3.47067
-3.24948
-3.06232
-2.90189
-2.76286
-2.64121
-2.53386
-2.43845
-2.35308
y2 = 
0.1
0.0909091
0.0833333
0.0769231
0.0714286
0.0666667
0.0625
0.0588235
0.0555556
0.0526316
0.05
0.047619
